File: leetcode.com/ransom-note.py
#!/bin/python3
# https://leetcode.com/problems/ransom-note/

import collections

class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        note_histogram = collections.Counter(ransomNote)
        magazine_histogram = collections.Counter(magazine)

        for c, count in note_histogram.items():
            if magazine_histogram[c] < count:
                return False

        return True

File: leetcode.com/ransom-note.rb
# @param {String} ransom_note
# @param {String} magazine
# @return {Boolean}
def can_construct(ransom_note, magazine)
    magazine_counter = {}

    magazine.each_char { |c| magazine_counter[c] = magazine_counter.key?(c) ? magazine_counter[c] + 1 : 1 }
    
    ransom_note.each_char do |c|
        return false if !magazine_counter.key?(c) || magazine_counter[c] == 0
        magazine_counter[c] -= 1
    end

    return true
end